system/inc/system_sleep.h

```diff
@@ -41,12 +51,406 @@ typedef enum System_Sleep_Flag
+    UNKNOWN = HAL_WAKEUP_SOURCE_TYPE_UNKNOWN,
+    BY_GPIO = HAL_WAKEUP_SOURCE_TYPE_GPIO,
+    BY_ADC = HAL_WAKEUP_SOURCE_TYPE_ADC,
+    BY_DAC = HAL_WAKEUP_SOURCE_TYPE_DAC,
+    BY_RTC = HAL_WAKEUP_SOURCE_TYPE_RTC,
+    BY_LPCOMP = HAL_WAKEUP_SOURCE_TYPE_LPCOMP,
+    BY_UART = HAL_WAKEUP_SOURCE_TYPE_UART,
+    BY_I2C = HAL_WAKEUP_SOURCE_TYPE_I2C,
+    BY_SPI = HAL_WAKEUP_SOURCE_TYPE_SPI,
+    BY_TIMER = HAL_WAKEUP_SOURCE_TYPE_TIMER,
+    BY_CAN = HAL_WAKEUP_SOURCE_TYPE_CAN,
+    BY_USB = HAL_WAKEUP_SOURCE_TYPE_USB,
+    BY_BLE = HAL_WAKEUP_SOURCE_TYPE_BLE,
+    BY_NFC = HAL_WAKEUP_SOURCE_TYPE_NFC,
+    BY_NETWORK = HAL_WAKEUP_SOURCE_TYPE_NETWORK,
+};
+
+class SystemSleepConfigurationHelper {
+public:
+    SystemSleepConfigurationHelper(const hal_sleep_config_t* config)
+        : config_(config) {
+    }
+
+    bool cloudDisconnectRequested() const {
+#if HAL_PLATFORM_CELLULAR
+        if (wakeupByNetworkInterface(NETWORK_INTERFACE_CELLULAR)) {
+            return false;
+        }
+#endif
+#if HAL_PLATFORM_WIFI
+        if (wakeupByNetworkInterface(NETWORK_INTERFACE_WIFI_STA)) {
+            return false;
+        }
+#endif
+#if HAL_PLATFORM_MESH
+        if (wakeupByNetworkInterface(NETWORK_INTERFACE_MESH)) {
+            return false;
+        }
+#endif
+#if HAL_PLATFORM_ETHERNET
+        if (wakeupByNetworkInterface(NETWORK_INTERFACE_ETHERNET)) {
+            return false;
+        }
+#endif
+        return true; 
+    }
+
+    bool wakeupByNetworkInterface(network_interface_index index) const {
+        auto wakeup = wakeupSourceFeatured(HAL_WAKEUP_SOURCE_TYPE_NETWORK);
+        while (wakeup) {
+            auto networkWakeup = reinterpret_cast<const hal_wakeup_source_network_t*>(wakeup);
+            if (networkWakeup->index == index) {
+                return true;
+            }
+            wakeup = wakeupSourceFeatured(HAL_WAKEUP_SOURCE_TYPE_NETWORK, wakeup->next);
+        }
+        return false;
+    }
+
+    SystemSleepWait sleepWait() const {
+        return static_cast<SystemSleepWait>(config_->wait);
+    }
+
+    SystemSleepMode sleepMode() const {
+        return static_cast<SystemSleepMode>(config_->mode);
+    }
+
+    hal_wakeup_source_base_t*wakeupSource() const {
+        return config_->wakeup_sources;
+    }
+
+    hal_wakeup_source_base_t* wakeupSourceFeatured(hal_wakeup_source_type_t type) const {
+        return wakeupSourceFeatured(type, config_->wakeup_sources);
+    }
+
+    hal_wakeup_source_base_t* wakeupSourceFeatured(hal_wakeup_source_type_t type, hal_wakeup_source_base_t* start) const {
+        if (!start) {
+            return nullptr;
+        }
+        while (start) {
+            if (start->type == type) {
+                return start;
+            }
+            start = start->next;
+        }
+        return nullptr;
+    }
+
+private:
+    const hal_sleep_config_t* config_;
+};
+
+class SystemSleepConfiguration: protected SystemSleepConfigurationHelper {
```

Comment:
It can be used by system layer potentially though. For example, the system may enter sleep mode automatically if `SYSTEM_MODE(AUTOMATIC)` is specified when system is in idle state. Besides, it would mess up the spark_wiring_system.h, as I'd like all sleep relevant stuff are resides in a dedicated header file, i.e. system_sleep.h.

Reply:
It may very well then use `wiring` `System.sleep()` or even directly call HAL. My main point is that this class is used only by wiring APIs, and hence should ideally be defined immediately where it's used.

Reply:
To maintain the compatibility, we've reworked the existing system sleep APIs, which now need to construct a `SystemSleepConfiguration` object and then invoke the new system sleep API. So this class is also used by the system layer.